# Artificial Intelligence & Machine Learning - Task 5

## Ensemble Learning, Model Optimization & Real-World ML Pipeline

**Internship Project – MainCrafts**

In [14]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier

from sklearn.metrics import  classification_report, accuracy_score,precision_score, recall_score, f1_score, confusion_matrix


## Load the Data Set

In [4]:
data = load_breast_cancer()
X = pd.DataFrame(data.data, columns=data.feature_names)
y = data.target

## Train-Test Split

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=42, stratify=y)

# Feature Scaling

In [7]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Baseline Model(Logistic Regression)

In [8]:
model_lr = LogisticRegression(max_iter=1000)
model_lr.fit(X_train_scaled, y_train)

y_pred_lr = model_lr.predict(X_test_scaled)
print(classification_report(y_test, y_pred_lr))

              precision    recall  f1-score   support

           0       0.98      0.98      0.98        42
           1       0.99      0.99      0.99        72

    accuracy                           0.98       114
   macro avg       0.98      0.98      0.98       114
weighted avg       0.98      0.98      0.98       114



## Random Forest Model

In [9]:
rf = RandomForestClassifier(random_state=42)
rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)
print(classification_report(y_test, y_pred_rf))

              precision    recall  f1-score   support

           0       0.95      0.93      0.94        42
           1       0.96      0.97      0.97        72

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114



# Gradient  Boosting Model

In [10]:
gb = GradientBoostingClassifier()   
gb.fit(X_train, y_train)

y_pred_gb = gb.predict(X_test)
print(classification_report(y_test, y_pred_gb))

              precision    recall  f1-score   support

           0       0.97      0.90      0.94        42
           1       0.95      0.99      0.97        72

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114



## Hyperparameter Tuning(GridSearchCV)

In [11]:
param_grid ={
    'n_estimators':[50,100],
    'max_depth':[3,5,7],
}
grid = GridSearchCV(RandomForestClassifier(),param_grid, cv = 3)
grid.fit(X_train, y_train)
print("Best parameters:", grid.best_params_)

Best parameters: {'max_depth': 7, 'n_estimators': 50}


## Model Comparison

# Step 1 - Train the Three models

In [12]:
# Create the models

logistic_model = LogisticRegression(max_iter=10000)

random_forest_model = RandomForestClassifier(
    random_state=42
)

gradient_boosting_model = GradientBoostingClassifier(
    random_state=42
)

## Step 2: Make Predicitions and Calculate Metrics

In [15]:
# Store the models in a dictionary

models = {
    "Logistic Regression": logistic_model,
    "Random Forest": random_forest_model,
    "Gradient Boosting": gradient_boosting_model
}

# Create a list to store results
results = []

# Train and evaluate each model
for name, model in models.items():

    # Train the model
    model.fit(X_train, y_train)

    # Make predictions
    y_pred = model.predict(X_test)

    # Calculate evaluation metrics
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)

    # Store results
    results.append({
        "Model": name,
        "Accuracy": accuracy,
        "Precision": precision,
        "Recall": recall,
        "F1-Score": f1
    })

## STEP 3: Create a Professional Comparison Table

In [16]:
# Create a DataFrame for model comparison

comparison_df = pd.DataFrame(results)

# Display results rounded to 4 decimal places
comparison_df.round(4)

,Model,Accuracy,Precision,Recall,F1-Score
0,Logistic Regression,0.9649,0.9595,0.9861,0.9726
1,Random Forest,0.9561,0.9589,0.9722,0.9655
2,Gradient Boosting,0.9561,0.9467,0.9861,0.9660


## Final Optimized Model

In [17]:
# Get the best Random Forest model from GridSearchCV
best_model = grid.best_estimator_

# Make predictions
y_pred_best = best_model.predict(X_test)

# Calculate evaluation metrics
best_accuracy = accuracy_score(y_test, y_pred_best)
best_precision = precision_score(y_test, y_pred_best)
best_recall = recall_score(y_test, y_pred_best)
best_f1 = f1_score(y_test, y_pred_best)

# Display the results
print("Optimized Random Forest Performance")
print("-" * 40)
print("Accuracy:", round(best_accuracy, 4))
print("Precision:", round(best_precision, 4))
print("Recall:", round(best_recall, 4))
print("F1-Score:", round(best_f1, 4))

Optimized Random Forest Performance
----------------------------------------
Accuracy: 0.9561
Precision: 0.9589
Recall: 0.9722
F1-Score: 0.9655


## Model Stability Analysis

In [18]:
# Perform 5-fold cross-validation for all models

models_cv = {
    "Logistic Regression": LogisticRegression(max_iter=10000),
    "Random Forest": RandomForestClassifier(random_state=42),
    "Gradient Boosting": GradientBoostingClassifier(random_state=42)
}

cv_results = []

for name, model in models_cv.items():

    scores = cross_val_score(
        model,
        X,
        y,
        cv=5,
        scoring="accuracy"
    )

    cv_results.append({
        "Model": name,
        "Mean CV Accuracy": scores.mean(),
        "Standard Deviation": scores.std()
    })

cv_df = pd.DataFrame(cv_results)

cv_df.round(4)

,Model,Mean CV Accuracy,Standard Deviation
0,Logistic Regression,0.9508,0.0180
1,Random Forest,0.9561,0.0228
2,Gradient Boosting,0.9631,0.0210


##  Final Model Selection

### Final Decision

The final model selection was based on performance, stability, and interpretability.

Logistic Regression achieved the highest test accuracy of 0.9649 and the highest F1-score of 0.9726. It also had the lowest cross-validation standard deviation of 0.0180, indicating the most consistent performance across different folds.

Gradient Boosting achieved the highest mean cross-validation accuracy of 0.9631, demonstrating strong overall performance. However, Logistic Regression provided better test performance and greater interpretability.

Random Forest was optimized using GridSearchCV, but the optimized model did not improve its test performance compared with the original Random Forest model.

Therefore, Logistic Regression was selected as the final model because it provided the best balance of strong test performance, high stability, and interpretability.

## Conclusion

Three machine learning models were compared: Logistic Regression, Random Forest, and Gradient Boosting.

The models were evaluated using Accuracy, Precision, Recall, and F1-score. Cross-validation was also performed to evaluate model stability.

GridSearchCV was used to optimize the Random Forest model and identify the best hyperparameters.

Based on performance, stability, and interpretability, Logistic Regression was selected as the final model.